src/plot_api/plot_api.js

```diff
@@ -242,6 +242,28 @@ exports.plot = function(gd, data, layout, config) {
             fullLayout._glcanvas
                 .attr('width', fullLayout.width)
                 .attr('height', fullLayout.height);
+
+            var regl = fullLayout._glcanvas.data()[0].regl;
+            if(regl) {
+                // Unfortunately, this can happen when relayouting to large
+                // width/height on some browsers.
+                if(fullLayout.width !== regl._gl.drawingBufferWidth ||
+                    fullLayout.height !== regl._gl.drawingBufferHeight
+                 ) {
+                    var msg = 'WebGL context buffer and canvas dimensions do not match due to browser/WebGL bug.';
+                    if(fullLayout._redrawFromWrongGlDimensions) {
+                        Lib.error(msg);
+                    } else {
+                        Lib.log(msg + ' Clearing graph and plotting again.');
+                        Plots.cleanPlot([], {}, gd._fullData, fullLayout, gd.calcdata);
+                        Plots.supplyDefaults(gd);
+                        fullLayout = gd._fullLayout;
+                        Plots.doCalcdata(gd);
+                        fullLayout._redrawFromWrongGlDimensions = 1;
+                        return drawFramework();
```

Comment:
Nice! Still a substantial win as we don't need to redo all the drawing.

So, there's still recursion here - but it's just back into this function. So instead of a `_redrawFromWrongGlDimensions` flag on `_fullLayout` (that we'd probably have to handle more carefully in principle due to `relinkPrivateKeys`) could we use a local var in the outer scope?

Reply:
Ok, done in https://github.com/plotly/plotly.js/pull/2939/commits/3879a41d94423b5fb73a7547213fc6bf4bdc1b0a

Reply:
one less private key to worry about when we'll :hocho: `relinkPrivateKeys`.